File: ML/LogReg/Python/LogReg.py
import os
import pandas as pd
import matplotlib.pyplot as plt
import sys
# sys.path.append(os.path.join(os.path.dirname(__file__), '..', 'data_prep'))
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn import preprocessing, set_config

from sklearn.datasets import load_iris

# path = os.path.dirname(os.path.dirname(__file__))


# df = pd.read_csv(os.path.join(path, 'data/train.csv'))
# df = cleaner_logreg(df)
# print(df.head())
# df.to_csv(os.path.join(path, "data/logreg_v5/test_nouvelles_variables.csv"), encoding='utf-8', index=False)

iris = load_iris()
X = iris.data          # (150, 4) -> features numériques
y = iris.target        # (150,)   -> labels 0, 1, 2
feature_names = iris.feature_names
target_names = iris.target_names

# Si tu veux un DataFrame complet pour explorer rapidement
df = pd.DataFrame(X, columns=feature_names)
df["species"] = pd.Categorical.from_codes(y, target_names)

print(df.head())
print(df.shape)

# X, y = df.drop(columns=["Survived", 'LastName', 'LastNameGroupSize', 'Sex_Alone']), df["Survived"]
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.5, random_state=42
)

#On standardise les données #Ajouté à l'étape 4
set_config(transform_output="pandas")
scaler = preprocessing.StandardScaler().fit(X_train)
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

#Entraînement du modèle
logit = LogisticRegression(max_iter=10000)

C = [0.1, 0.25, 0.5, 0.75, 1, 1.25, 1.5, 1.75, 2]
scores = []
for choice in C:
    logit.set_params(C=choice)
    logit.fit(X_train, y_train)
    scores.append(logit.score(X_test, y_test))
print(scores)
print(f"Score test interne: {logit.score(X_test, y_test):.4f}")

#Test du modèle
# teste = pd.read_csv(os.path.join(path, "data/test.csv"))
# teste = cleaner_logreg(teste)
# teste = teste.drop(columns=['LastName', 'LastNameGroupSize', 'Sex_Alone'])
# teste['Deck_T'] = 0
# teste = teste[X_train.columns]#Pour avoir le même ordre dans les colonnes
# #On standardise les données #Ajouté à l'étape 4
# teste[['Fare', 'Age', 'SibSp', 'Parch', 'Family_size', 'TicketGroupSize']] = scaler.transform(teste[['Fare', 'Age', 'SibSp', 'Parch', 'Family_size', 'TicketGroupSize']])

logit.set_params(C=C[scores.index(max(scores))])
logit.fit(X_train, y_train)
y_best_pred = logit.predict(X_test)

# teste.to_csv(os.path.join(path, "data/logreg_v5/submission_logreg.csv"), encoding='utf-8', index=False)
plt.scatter(X[:, 2], y)


#Plot multinomial
import numpy as np

# On fait varier la longueur des pétales sur toute sa plage,
# en figeant les 3 autres features à leur moyenne (sur les données brutes)
x_range = np.linspace(X[:, 2].min(), X[:, 2].max(), 300)

X_synth = np.tile(X.mean(axis=0), (len(x_range), 1))  # autres features = moyenne
X_synth[:, 2] = x_range                                 # on ne fait varier que la colonne 2

X_synth_df = pd.DataFrame(X_synth, columns=feature_names)
X_synth_scaled = scaler.transform(X_synth_df)           # même standardisation que l'entraînement

probas = logit.predict_proba(X_synth_scaled)            # shape (300, 3)

# Tracé : une courbe en S par classe
plt.figure(figsize=(8, 5))
for i, name in enumerate(target_names):
    plt.plot(x_range, probas[:, i], label=f"P({name})")



#Plot binomial
# Points observés : on les place à 0 ou 1 sur la classe correspondante,
# avec un peu de bruit vertical pour la lisibilité (jitter)
for i in range(3):
    mask = (y == i)
    jitter = np.random.uniform(-0.02, 0.02, mask.sum())
    plt.scatter(X[mask, 2], np.ones(mask.sum()) + jitter - 1 + i*0,
                alpha=0.3, s=15, color=f"C{i}")

plt.xlabel("Longueur des pétales (cm)")
plt.ylabel("Probabilité prédite")
plt.title("Probabilités par classe vs longueur des pétales")
plt.legend()
plt.show()


y_binary = (y == 0).astype(int)  # setosa = 1, les autres = 0
logit_demo = LogisticRegression().fit(X[:, [2]], y_binary)  # 1 seule feature, non standardisée pour rester lisible

x_range = np.linspace(X[:, 2].min(), X[:, 2].max(), 300).reshape(-1, 1)
proba_demo = logit_demo.predict_proba(x_range)[:, 1]

plt.scatter(X[:, 2], y_binary, alpha=0.5)
plt.plot(x_range, proba_demo, color="red")
plt.xlabel("Longueur des pétales (cm)")
plt.ylabel("P(setosa)")
plt.title("Régression logistique binaire - courbe sigmoïde")
plt.show()

File: ML/LogReg/R/LogReg.R
# install.packages("glmnet")
library(glmnet)

## ============================================================
## 1. Données
## ============================================================
data(iris)
X <- as.matrix(iris[, 1:4])
y <- iris$Species
feature_names <- colnames(X)
target_names  <- levels(y)

print(head(iris))
print(dim(iris))

## ============================================================
## 2. Split train/test 50/50 (équivalent random_state=42)
## ============================================================
set.seed(42)
n <- nrow(X)
train_idx <- sample(seq_len(n), size = floor(0.5 * n))

X_train <- X[train_idx, ];  X_test <- X[-train_idx, ]
y_train <- y[train_idx];    y_test <- y[-train_idx]

## ============================================================
## 3. Standardisation (moyenne/sd calculées sur train uniquement)
## ============================================================
train_means <- colMeans(X_train)
train_sds   <- apply(X_train, 2, sd)

scale_with <- function(mat, means, sds) sweep(sweep(mat, 2, means, "-"), 2, sds, "/")

X_train_s <- scale_with(X_train, train_means, train_sds)
X_test_s  <- scale_with(X_test,  train_means, train_sds)

## ============================================================
## 4. Calcul du chemin complet de régularisation
##    (au lieu d'un lambda isolé -> warm start correct, coefficients fiables)
## ============================================================
full_path <- glmnet(X_train_s, y_train, family = "multinomial", alpha = 0)

# Scores sur tout le chemin
pred_class_path <- predict(full_path, X_test_s, type = "class")          # matrice (n_test x n_lambda)
scores_path <- apply(pred_class_path, 2, function(col) mean(col == y_test))

# Sélection du meilleur lambda (premier max en cas d'égalité)
best_idx    <- which.max(scores_path)
best_lambda <- full_path$lambda[best_idx]
cat(sprintf("Meilleur lambda : %.5f (score test = %.4f)\n",
            best_lambda, scores_path[best_idx]))

# Pour information : C approximatif correspondant (1/lambda, cf. avertissement précédent)
cat(sprintf("C équivalent approx. : %.4f\n", 1 / best_lambda))

## ============================================================
## 5. Prédiction finale au meilleur lambda
## ============================================================
y_best_pred <- predict(full_path, X_test_s, type = "class", s = best_lambda)
cat(sprintf("Score test interne : %.4f\n", mean(y_best_pred == y_test)))

## ============================================================
## 6. Scatter brut : longueur des pétales vs classe réelle
## ============================================================
y_numeric <- as.numeric(y) - 1
plot(X[, 3], y_numeric, xlab = "Longueur des pétales (cm)", ylab = "Classe",
     main = "Longueur des pétales vs classe", pch = 16,
     col = as.numeric(y))

## ============================================================
## 7. Plot multinomial : probabilités prédites par classe
## ============================================================
x_range <- seq(min(X[, 3]), max(X[, 3]), length.out = 300)

X_synth <- matrix(rep(colMeans(X), 300), nrow = 300, byrow = TRUE)
X_synth[, 3] <- x_range
colnames(X_synth) <- feature_names

X_synth_s <- scale_with(X_synth, train_means, train_sds)

# Prédiction au bon lambda via s= (et non plus un fit isolé)
probas <- predict(full_path, X_synth_s, type = "response", s = best_lambda)[, , 1]  # 300 x 3

plot(x_range, probas[, 1], type = "l", col = 1, lwd = 2,
     xlab = "Longueur des pétales (cm)", ylab = "Probabilité prédite",
     main = "Probabilités par classe vs longueur des pétales",
     ylim = c(0, 1))
lines(x_range, probas[, 2], col = 2, lwd = 2)
lines(x_range, probas[, 3], col = 3, lwd = 2)

# Points observés, alignés sur leur vraie classe (0, 1 ou 2) avec un jitter vertical
for (i in seq_along(target_names)) {
  mask   <- (y_numeric == (i - 1))
  jitter <- runif(sum(mask), -0.02, 0.02)
  points(X[mask, 3], rep((i - 1) / 2, sum(mask)) + jitter,  # repère visuel par classe, pas une vraie probabilité
         col = i, pch = 16, cex = 0.6)
}
legend("topleft", legend = paste0("P(", target_names, ")"), col = 1:3, lwd = 2)

## ============================================================
## 8. Plot binomial : setosa vs reste, courbe sigmoïde (glm, non régularisé)
## ============================================================
y_binary <- as.integer(y == "setosa")

logit_demo <- glm(y_binary ~ petal_length,
                  data = data.frame(y_binary = y_binary, petal_length = X[, 3]),
                  family = binomial)

x_range2   <- data.frame(petal_length = seq(min(X[, 3]), max(X[, 3]), length.out = 300))
proba_demo <- predict(logit_demo, newdata = x_range2, type = "response")

plot(X[, 3], y_binary, xlab = "Longueur des pétales (cm)", ylab = "P(setosa)",
     main = "Régression logistique binaire - courbe sigmoïde", pch = 16,
     col = rgb(0, 0, 0, 0.4))
lines(x_range2$petal_length, proba_demo, col = "red", lwd = 2)